tracepoint_decode/src/perf_event_data.rs

```diff
@@ -0,0 +1,583 @@
+    ///     decoded uses the event's `format` (if available).
+    ///
+    /// You will normally use `reader.get_sample_event_info(&bytes)` to help parse the
+    /// fields from the data of the `Sample` event.
+    ///
+    /// If this is a non-sample event (i.e. if `header.header_type != PerfEventHeaderType::Sample`)
+    /// then `data` will contain:
+    ///
+    /// - The 8 byte header (same data as `header`, but in event byte order).
+    /// - The content of the event (if any), which is in a format determined by the event's
+    ///   `header_type`.
+    /// - The event may contain additional fields after the content, one field for each bit
+    ///   set in the event's `sample_type`. These fields include items like `id`, `time`,
+    ///   `cpu`, pid, tid.
+    ///
+    /// If additional fields are present, you will normally use
+    /// `reader.get_non_sample_event_info(&bytes)` to parse them. Note that these addtional
+    /// fields are not always present. In particular, they are generally not present for
+    /// events with `header_type >= UserTypeStart` or for events that appear before the
+    /// `FinishedInit` event has been processed.
+    pub data: &'dat [u8],
+}
+
+impl<'dat> PerfEventBytes<'dat> {
+    /// Constructs a new PerfEventBytes instance.
+    pub const fn new(header: PerfEventHeader, data: &'dat [u8]) -> PerfEventBytes<'dat> {
+        debug_assert!(data.len() >= mem::size_of::<PerfEventHeader>());
+        return PerfEventBytes { header, data };
+    }
+}
+
+/// Information about a non-sample event, typically returned by
+/// `reader.get_non_sample_event_info(&bytes)`.
+#[derive(Clone, Debug)]
+pub struct PerfNonSampleEventInfo<'a> {
+    /// The bytes of the event, including header and content, in event byte order.
+    ///
+    /// The data consists of the 8-byte header followed by the content, both in event byte order.
+    /// The format of the content depends on `header_type`.
+    ///
+    /// Valid always.
+    pub data: &'a [u8],
+
+    /// Information about the session that collected the event, e.g. clock id and
+    /// clock offset.
+    ///
+    /// Valid always.
+    pub session_info: &'a PerfSessionInfo,
+
+    /// Information about the event (shared by all events with the same `id`).
+    ///
+    /// Valid always.
+    pub event_desc: &'a PerfEventDesc,
+
+    /// Valid if `sample_type()` contains `Identifier` or `Id`.
+    pub id: u64,
+
+    /// Valid if `sample_type()` contains `Cpu`.
+    pub cpu: u32,
+
+    /// Valid if `sample_type()` contains `Cpu`.
+    pub cpu_reserved: u32,
+
+    /// Valid if `sample_type()` contains `StreamId`.
+    pub stream_id: u64,
+
+    /// Use SessionInfo.TimeToTimeSpec() to convert to a TimeSpec.
+    ///
+    /// Valid if `sample_type()` contains `Time`.
+    pub time: u64,
+
+    /// Valid if `sample_type()` contains `Tid`.
+    pub pid: u32,
+
+    /// Valid if `sample_type()` contains `Tid`.
+    pub tid: u32,
```

Comment:
Instead of "valid if xyz" in the comments, these should probably be accessors that return an Option<>